contracts/cw3-flex-multisig/src/contract.rs

```diff
@@ -468,230 +454,285 @@ mod tests {
-            e => panic!("unexpected error: {}", e),
-        }
+        assert_eq!(res.unwrap_err(), ContractError::ZeroWeight {}.to_string());
 
         // Total weight less than required weight not allowed
-        let required_weight = 100;
-        let res = setup_test_case(
-            deps.as_mut(),
-            info.clone(),
-            required_weight,
+        let init_msg = InitMsg {
+            group: group_addr.clone(),
+            required_weight: 100,
             max_voting_period,
-        );
+        };
+        let res = app.instantiate_contract(flex_id, OWNER, &init_msg, &[], "high required weight");
 
         // Verify
-        assert!(res.is_err());
-        match res.unwrap_err() {
-            ContractError::UnreachableWeight {} => {}
-            e => panic!("unexpected error: {}", e),
-        }
+        assert_eq!(
+            res.unwrap_err(),
+            ContractError::UnreachableWeight {}.to_string()
+        );
 
         // All valid
-        let required_weight = 1;
-        setup_test_case(deps.as_mut(), info, required_weight, max_voting_period).unwrap();
+        let init_msg = InitMsg {
+            group: group_addr.clone(),
+            required_weight: 1,
+            max_voting_period,
+        };
+        let flex_addr = app
+            .instantiate_contract(flex_id, OWNER, &init_msg, &[], "all good")
+            .unwrap();
 
-        // Verify
+        // Verify contract version set properly
+        let version = query_contract_info(&app, &flex_addr).unwrap();
         assert_eq!(
             ContractVersion {
                 contract: CONTRACT_NAME.to_string(),
                 version: CONTRACT_VERSION.to_string(),
             },
-            get_contract_version(&deps.storage).unwrap()
-        )
+            version,
+        );
+
+        // Get voters query
+        let voters: VoterListResponse = app
+            .wrap()
+            .query_wasm_smart(
+                &flex_addr,
+                &QueryMsg::ListVoters {
+                    start_after: None,
+                    limit: None,
+                },
+            )
+            .unwrap();
+        assert_eq!(
+            voters.voters,
+            vec![VoterResponse {
+                addr: OWNER.into(),
+                weight: 1
+            }]
+        );
     }
 
     // TODO: query() tests
 
     #[test]
     fn test_propose_works() {
-        let mut deps = mock_dependencies(&[]);
+        let mut app = mock_app();
 
         let required_weight = 4;
         let voting_period = Duration::Time(2000000);
-
-        let info = mock_info(OWNER, &[]);
-        setup_test_case(deps.as_mut(), info.clone(), required_weight, voting_period).unwrap();
+        let flex_addr = setup_test_case(&mut app, required_weight, voting_period, coins(10, "BTC"));
 
         let bank_msg = BankMsg::Send {
-            from_address: OWNER.into(),
+            from_address: flex_addr.clone(),
```

Comment:
If I understand correctly, this is not (not necessarily) wrong in the original (fixed-multisig) code. It's just another way to reward the recipient. Better, in a sense, as the contract must have the funds already assigned, instead of depending on a third party for them.

I wonder what would happen if a proposal is passed, contract releases the bank messages, but the sender (a third party, or the contract iself) lacks enough funds.

A sanity check in multisig contracts code, only accepting proposals if the contract is funded enough to eventually release the associated bank messages, would be a good idea here. Of course, I think this is only possible in the current scenario, that is, when the contract addr is in the `from_addr` of the `BankMsg`s.

Reply:
If you are OK with it, I'll create an issue to add this check to both multisig contracts.

Reply:
If the contract passes a proposal to send from the owners account, the bank module will reject it.

I wanted to remove the from field after some confusion as it is currently only supported when from is the contract itself (if it can move funds from your account, that is a major security hole).

That is a bug in the fixed contract, I can fix that as well

Reply:
It is very hard in general to validate if any message will pass. Even more if it will be valid at the end of the voting period.

I would not do any such checks in the contract, but encourage any UI to do so

Reply:
Currently what would happen is execute fails. If it was sent from the contract and there was insufficient funds, you could try to execute later if more funds arrive

Reply:
OK then. You're right, that is a bug / confusion in the original multisig test code.

Reply:
> I wanted to remove the from field after some confusion as it is currently only supported when from is the contract itself (if it can move funds from your account, that is a major security hole).

I think that is a good idea. It will remove a whole class of bugs, and simplify setup for use cases, tests, etc.

In fact, allowing a message payload in the handle functions is probably not a good idea in general. Better for the contract to build the messages, and pass only the required / variable fields to the handle.

Reply:
> (if it can move funds from your account, that is a major security hole).

Of course. I get confused between the owner / creator / admin address, and the contract address itself.